### ISSUE TITLE
Verify email with password

### ISSUE THREAD

**Issue by 9mido** [pennersr/django-allauth#2699]
TLDR - ask users for their password to verify their email address. This ensures additional account added unverified email addresses associated with a user cannot login with an unverified email address and lock out the real user from their account unless they know the user's password. We will also need prevent unverified users from requesting password resets to avoid account lockouts. 

Scenario:

1. User A owns a@a.com
2. User A adds mistyped email b@b.com
3. User B owns b@b.com, receives confirmation email
4. User B cannot login because the confirmation page has a password requirement (thanks to the code sample below)
5. User B does a password reset for b@b.com 
6. User B can login with the reset password and lock out user A 

To solve 5-6, we need to only allow verified users to be able to complete password resets. See #2360. I have written a solution for this in 2360. 

If someone makes a PR for this, we will need the following settings:
1. ACCOUNT_PASSWORD_RESET_VERIFIED_ONLY=True|False
2. ACCOUNT_EMAIL_CONFIRMATION_REQUIRES_PASSWORD=True|False

Related to now closed PR #2627 (see for more details)

When changing email addresses on Stripe's website, their email address verification process has a password requirement when clicking the emailed verification link to verify the email address successfully and to ensure that the email account is actually owned by the true owner. 

This password requirement for verification should be for user signups and changing email addresses.

For signups, the password would need to be entered on the /accounts/confirm-email/HASH_GOES_HERE page to complete the verification after the user clicks the /accounts/confirm-email/HASH_GOES_HERE URL sent in the verification email. Add a password field to the /accounts/confirm-email/HASH_GOES_HERE page and have the user type in their password then press submit to confirm and authorize. If the password is correct, complete the verification. The email address should have already been gathered from the email that was sent. If not, the email address will already be added in the database before the user confirms/completes the verification.

The same thing would be need to be done for /accounts/email/ page when adding/changing a new email address / re-sending verification since it is the same process as signups since you land on /accounts/confirm-email/HASH_GOES_HERE both times.

Since this will be another form element, another entry will probably be needed for ACCOUNT_FORMS = 'verification':'app_name.forms.EmailVerificationForm' as an example to allow the django developer to customize the verification password form. 

Or maybe we don't need to have a setting and could just add this in as a default.

Maybe this can help someone make a PR for verifying email with password:
https://stackoverflow.com/questions/4822724/check-password-from-a-user-again

**Comment by 9mido** [comment]
Made some progress. If anyone could make a PR for this, please do and feel free to make changes. This is how I solved it:

email_confirm.html:

`{{ form.as_p }}`

Forms.py (put this code after the PasswordField class):

```
class EmailConfirmForm(forms.Form):
    password = PasswordField(label=_("Password"))

    def __init__(self, user, data, *args, **kwargs):
        self.user = user
        super(EmailConfirmForm, self).__init__(data=data, *args, **kwargs)
        self.fields['password'].widget.attrs['placeholder'] = 'Password'
```

Views.py (put this code in the ConfirmEmailView class):

from .forms import (EmailConfirmForm,)
from django.http import HttpResponseServerError
from django.contrib import messages

       # Inside the post function before the "if app_settings.LOGIN_ON_EMAIL_CONFIRMATION" add this code:
        user = confirmation.email_address.user
        success = user.check_password(self.request.POST['password'])
        if not success:
            return HttpResponseServerError("wrong password")

        # Inside the get_context_data function add this code:
        confirmation = self.get_object()
        user = confirmation.email_address.user_id
        form = EmailConfirmForm(user=confirmation.email_address.user, data=self.request.POST)
        ctx.update({"site": site, "form":form,})

**Comment by pennersr** [comment]
How would you deal with this when the account does not have a password set (i.e. a social account) ? Basically, the requirement would be not to verify email with a password, but verify email by being logged in.

**Comment by 29deepanshutyagi** [comment]
i want to work on this issue, kindly assign me this issue

**Comment by pennersr** [comment]
@29deepanshutyagi This issue is not clear to me. It tries to prevent a scenario that does not make much sense to me. And, the suggested solution does not work in all circumstances, e.g. not all accounts have a password. So, I don't think this is a valid feature request.

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/2699
